tls/s2n_kem.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "stuffer/s2n_stuffer.h"
+
+#include "tls/s2n_kem.h"
+
+#include "utils/s2n_safety.h"
+#include "utils/s2n_mem.h"
+
+int s2n_kem_generate_keypair(struct s2n_kem_keypair *kem_keys)
+{
+    const struct s2n_kem *kem = kem_keys->negotiated_kem;
+    notnull_check(kem->generate_keypair);
+
+    eq_check(kem_keys->public_key.size, kem->public_key_length);
+    notnull_check(kem_keys->public_key.data);
+
+    // The private key is needed for client_key_recv and must be saved
+    GUARD(s2n_alloc(&kem_keys->private_key, kem->private_key_length));
+
+    GUARD(kem->generate_keypair(kem_keys->public_key.data, kem_keys->private_key.data));
+    return 0;
+}
+
+int s2n_kem_encapsulate(const struct s2n_kem_keypair *kem_keys, struct s2n_blob *shared_secret,
```

Comment:
Are we really encapsulating here? Reading through it seems more like we're deriving the shared secret?

Reply:
Its both, the client uses a KEM to generates the shared secret and encapsulates it with server's public key. The server can decapsulate the shared secret with the private key and ciphertext.

I talked with Nir about the names and we decided to have the s2n wrapper methods mirror the NIST API from https://csrc.nist.gov/csrc/media/projects/post-quantum-cryptography/documents/example-files/api-notes.pdf like this. Something like `s2n_kem_generate_shared_secret_and_encapsulate` didn't roll off the tongue.